lib/streamlit/elements/dialog_decorator.py

```diff
@@ -0,0 +1,172 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import contextlib
+import inspect
+from functools import wraps
+from typing import Callable, TypeVar, cast, overload
+
+from streamlit.delta_generator import event_dg, get_last_dg_added_to_context_stack
+from streamlit.elements.lib.dialog import DialogWidth
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.fragment import fragment as _fragment
+
+
+def _assert_no_nested_dialogs() -> None:
+    """Check the current stack for existing DeltaGenerator's of type 'dialog'.
+    Note that the check like this only works when Dialog is called as a context manager, as this populates the dg_stack in delta_generator correctly.
+
+    This does not detect the edge case in which someone calls, for example, `with st.sidebar` inside of a dialog function and opens a dialog in there,
+    as `with st.sidebar` pushes the new DeltaGenerator to the stack. In order to check for that edge case, we could try to check all DeltaGenerators in the stack,
+    and not only the last one. Since we deem this to be an edge case, we lean towards simplicity here.
+
+    Raises
+    ------
+    StreamlitAPIException
+        Raised if the user tries to nest dialogs inside of each other.
+    """
+    last_dg_in_current_context = get_last_dg_added_to_context_stack()
+    if last_dg_in_current_context and "dialog" in set(
+        last_dg_in_current_context._ancestor_block_types
+    ):
+        raise StreamlitAPIException("Dialogs may not be nested inside other dialogs.")
+
+
+F = TypeVar("F", bound=Callable[..., None])
+
+
+def _dialog_decorator(
+    non_optional_func: F, title: str, *, width: DialogWidth = "small"
+) -> F:
+    if title is None or title == "":
+        raise StreamlitAPIException(
+            'A non-empty `title` argument has to be provided for dialogs, for example `@st.experimental_dialog("Example Title")`.'
+        )
+
+    @wraps(non_optional_func)
+    def wrap(*args, **kwargs) -> None:
+        _assert_no_nested_dialogs()
+        # Call the Dialog on the event_dg because it lives outside of the normal
+        # Streamlit UI flow. For example, if it is called from the sidebar, it should not
+        # inherit the sidebar theming.
+        dialog = event_dg.dialog(title=title, dismissible=True, width=width)
+        dialog.open()
+
+        @_fragment
+        def dialog_content() -> None:
+            # if the dialog should be closed, st.rerun() has to be called (same behavior as with st.fragment)
+            _ = non_optional_func(*args, **kwargs)
+            return None
+
+        with dialog:
+            return dialog_content()
+
+    with contextlib.suppress(AttributeError):
+        # Make this a well-behaved decorator by preserving important function
+        # attributes.
```

Comment:
I have shamelessly stolen this from @vdonato's fragment code ([here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/runtime/fragment.py#L304-L308)), so I thought its best-practice. From my understanding it is for preserving the wrapped function's signature etc., e.g. when errors are thrown or when you call `help(...)` on the wrapped function (if yes, I can update the comment to reflect that better).
However, I have just done a little bit more research and it sounds like that `@wraps` is already doing this since Python 3.4+ and might only be needed for previous versions. Afaik we don't support versions <3.8 anymore though, right? So we might be able to get rid of those lines completely. 
@vdonato what do you think? is that understanding correct?

Reply:
Ah hm, I also more or less added it routinely because I thought it was necessary, but if `@wraps` does this automatically in every version of Python we need to support, we can probably go ahead and get rid of this.